rust/src/shards/gui/layouts/table.rs

```diff
@@ -0,0 +1,429 @@
+      let mut should_expose = true;
+      for var in &shared {
+        let (a, b) = unsafe {
+          (
+            CStr::from_ptr(var.name),
+            CStr::from_ptr(self.row_index.get_name()),
+          )
+        };
+        if CStr::cmp(a, b) == Ordering::Equal {
+          should_expose = false;
+          if var.exposedType.basicType != SHType_Int {
+            return Err("Table: int variable required.");
+          }
+          break;
+        }
+      }
+
+      if should_expose {
+        // expose row index
+        let index_info = ExposedInfo {
+          exposedType: common_type::int,
+          name: self.row_index.get_name(),
+          help: cstr!("The row index.").into(),
+          isMutable: false,
+          isProtected: false,
+          isTableEntry: false,
+          global: false,
+          isPushTable: false,
+        };
+        shared.push(index_info);
+        // update shared
+        data.shared = (&shared).into();
+      }
+    }
+
+    let input_type = data.inputType;
+    let slice = unsafe {
+      let ptr = input_type.details.seqTypes.elements;
+      std::slice::from_raw_parts(ptr, input_type.details.seqTypes.len as usize)
+    };
+    let element_type = if !slice.is_empty() {
+      slice[0]
+    } else {
+      common_type::any
+    };
+
+    for s in &mut self.shards {
+      data.inputType = element_type;
+      s.compose(&data)?;
+    }
+
+    // Always passthrough the input
+    Ok(input_type)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+
+    self.rows.warmup(ctx);
+    self.columns.warmup(ctx);
+    self.striped.warmup(ctx);
+    self.resizable.warmup(ctx);
+    self.row_index.warmup(ctx);
+
+    if self.row_index.is_variable() {
+      self.row_index.get_mut().valueType = common_type::int.basicType;
+    }
+
+    for s in &mut self.shards {
+      s.warmup(ctx)?;
+    }
+    for s in &mut self.header_shards {
+      if let Some(s) = s {
+        s.warmup(ctx)?;
+      }
+    }
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    for s in &mut self.header_shards {
+      if let Some(s) = s {
+        s.cleanup();
+      }
+    }
+    for s in &mut self.shards {
+      s.cleanup();
+    }
+    self.row_index.cleanup();
+    self.resizable.cleanup();
+    self.striped.cleanup();
+    self.columns.cleanup();
+    self.rows.cleanup();
+
+    self.parents.cleanup();
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      ui.push_id(EguiId::new(self, 0), |ui| {
+        use egui_extras::{Size, TableBuilder};
+
+        let seq = Seq::try_from(input)?;
+        let text_height = egui::TextStyle::Body.resolve(ui.style()).size;
+        // start building a table
+        let mut builder =
+          TableBuilder::new(ui).cell_layout(egui::Layout::left_to_right(egui::Align::Center));
```

Comment:
applies to all egui but within Shards all those can be easily cached

Reply:
The way egui works, most of the time they can't be cached: they use the "builder pattern" which means they return a moved `Self`, not a reference. So after being cached, nothing can be changed again (some calls can't be made anymore because of moving `Self`).